layer/layer.go

```diff
@@ -107,6 +108,14 @@ type Layer interface {
 	Metadata() (map[string]string, error)
 }
 
+// ForeignSourcer is an interface used to describe the source of layers
+// and objects representing layers, when the source is a foreign URL.
+type ForeignSourcer interface {
+	// ForeignSource returns the descriptor for this layer if it is
+	// a foreign layer, or nil for ordinary layers.
+	ForeignSource() *distribution.Descriptor
```

Comment:
so rather than returning the struct we do return the interface, is this what you mean? fine however